server/form-pages/apply/risk-and-need-factors/oasys-import/optionalOasysSections.ts

```diff
@@ -0,0 +1,115 @@
+import { DataServices } from '../../../../@types/ui'
+import { sentenceCase } from '../../../../utils/utils'
+
+interface Response {
+  needsLinkedToReoffending: string[] | string | OASysSection[]
+  otherNeeds: string[] | string | OASysSection[]
+}
+
+interface Body {
+  needsLinkedToReoffending: string[]
+  otherNeeds: string[]
+}
+
+@Page({
+  name: 'optional-oasys-sections',
+  bodyProperties: ['needsLinkedToReoffending', 'otherNeeds'],
+})
+export default class OptionalOasysSections implements TasklistPage {
+  title = 'Which of the following sections of OASys do you want to import?'
+
+  needsLinkedToReoffendingHeading = 'Needs linked to reoffending'
+
+  allNeedsLinkedToReoffending: OASysSection[]
+
+  otherNeedsHeading = 'Needs not linked to risk of serious harm or reoffending'
+
+  allOtherNeeds: OASysSection[]
+
+  constructor(public body: Partial<Body>) {}
+
+  static async initialize(
+    body: Partial<Response>,
+    application: Application,
+    token: string,
+    dataServices: DataServices,
+  ) {
+    const oasysSelections = await dataServices.personService.getOasysSelections(token, application.person.crn)
+
+    const allNeedsLinkedToReoffending = oasysSelections.filter(
+      section => !section.linkedToHarm && section.linkedToReOffending,
+    )
+    const allOtherNeeds = oasysSelections.filter(section => !section.linkedToHarm && !section.linkedToReOffending)
+
+    const lists = ['needsLinkedToReoffending', 'otherNeeds']
+
+    lists.forEach(section => {
+      if (isStringOrArrayOfStrings(body[section])) {
+        body[section] = flattenCheckboxInput(body[section])
+      }
+    })
+
+    const page = new OptionalOasysSections(body as Body)
+
+    page.allNeedsLinkedToReoffending = allNeedsLinkedToReoffending
+    page.allOtherNeeds = allOtherNeeds
+
+    return page
+  }
+
+  previous() {
+    return ''
+  }
+
+  next() {
+    return ''
+  }
+
+  response() {
+    const response = {}
+
+    if (this.getResponseForTypeOfNeed(this.body.needsLinkedToReoffending))
+      response[this.needsLinkedToReoffendingHeading] = this.getResponseForTypeOfNeed(this.body.needsLinkedToReoffending)
+
+    if (this.getResponseForTypeOfNeed(this.body.otherNeeds))
+      response[this.otherNeedsHeading] = this.getResponseForTypeOfNeed(this.body.otherNeeds)
+
+    return response
+  }
+
+  getResponseForTypeOfNeed(typeOfNeed: string[]) {
+    if (Array.isArray(typeOfNeed) && typeOfNeed.length) {
+      return typeOfNeed.join(', ')
+    }
+    return ''
+  }
+
+  reoffendingNeedsItems() {
+    return this.checkboxes(this.allNeedsLinkedToReoffending, this.body.needsLinkedToReoffending)
+  }
+
+  otherNeedsItems() {
+    return this.checkboxes(this.allOtherNeeds, this.body.otherNeeds)
+  }
+
+  private checkboxes(fullList: OASysSection[], selectedList: string[]) {
+    return fullList.map(need => {
+      const sectionAndName = `${need.section}. ${sentenceCase(need.name)}`
+      return {
+        value: sectionAndName,
```

Comment:
Should the value be just the section number? It'll make it tricky to use in the next step if we have to extract the section number

Reply:
I was having trouble with this as we then had to look up the section name when doing the CYA pages. It's solvable but I will need a bit of guidance

Reply:
Could we not do the lookup at the point of submission and then store the response as a key/pair value by overriding the body setter?